classes/admin/saml2_settings.php

```diff
@@ -41,4 +41,18 @@ abstract class saml2_settings {
     const OPTION_MULTI_IDP_DISPLAY_DROPDOWN = 0;
 
     const OPTION_MULTI_IDP_DISPLAY_BUTTONS = 1;
+
+    const OPTION_FLAGGED_LOGIN_MESSAGE = 1;
+
+    const OPTION_FLAGGED_LOGIN_REDIRECT = 2;
+
+    /**
+     * @var string representation of PCRE Regular Expression for checking http/https scheme URLs
+     * in accordance with RFC3986.
+     */
+    const URL_HTTPS_REGEX = '/(^(https?\\:\\/\\/(www\\.)?'
```

Comment:
This is gross. I read the comment about SAML lib being more picky than Moodle lib, but I still don't understand why we need it. It appears it's Moodle core that processes this, or did I miss something?

Reply:
Actually it could be a leftover from the initial implementation. I'll check the code. @tomdickman can you please remind us why it was done in this way?

Reply:
Hey @mattporritt  and @dmitriim, we had a discussion about this originally as I didn't really want to have to do it and instead I wanted to use the Moodle inbuilt form param types for adding the setting, However, in testing none of the param constants worked for URLs as expected. Specifically `PARAM_URL` still let through poorly formed URLs (in fact it didn't seem to parse anything at all if I remember correctly, it's been a while).
As a result the decision was made to use regex and just add some extensive test coverage, as a workaround.

Reply:
@mattporritt it's up to you to keep this code or I can just use PARAM_URL and leave it to admins to make sure that not screw it up.

Reply:
Replaced with PARAM_URL and removed unused code.